internal/mode/static/config_updater.go

```diff
@@ -53,18 +74,64 @@ func (z zapSetterImpl) SetLevel(level string) error {
 }
 
 // Enabled returns true if the given level is at or above the current level.
-func (z zapSetterImpl) Enabled(level zapcore.Level) bool {
+func (z zapLogLevelSetter) Enabled(level zapcore.Level) bool {
 	return z.atomicLevel.Enabled(level)
 }
 
+// leveledPrometheusLogger is a leveled prometheus logger.
+// This interface is required because the promlog.NewDynamic returns an unexported type *logger.
+type leveledPrometheusLogger interface {
+	log.Logger
+	SetLevel(level *promlog.AllowedLevel)
+}
+
+type promLogLevelSetter struct {
+	logger leveledPrometheusLogger
+}
+
+func newPromLogLevelSetter(logger leveledPrometheusLogger) promLogLevelSetter {
+	return promLogLevelSetter{logger: logger}
+}
+
+func newLeveledPrometheusLogger() (leveledPrometheusLogger, error) {
+	logFormat := &promlog.AllowedFormat{}
+
+	if err := logFormat.Set("json"); err != nil {
+		return nil, err
+	}
+
+	logConfig := &promlog.Config{Format: logFormat}
+	logger := promlog.NewDynamic(logConfig)
+
+	return logger, nil
+}
+
+func (p promLogLevelSetter) SetLevel(level string) error {
+	al := &promlog.AllowedLevel{}
+	if err := al.Set(level); err != nil {
+		fieldErr := field.NotSupported(
```

Comment:
I think it'd be better if we validate before we call SetLevel. I don't think levelSetters should not have knowledge about how to format errors (like call field.NotSupported) or what field is being used (here is it is `logging.level`).

Could we perform validation for fields of ControlPlane config in `updateControlPlane` function?

In SetLevel, we can still return errors, but they could be much simpler - we don't need to worry about field names and stuff. the details about valid values here can be taken by wrapping the error from the corresponding logger methods.